Fix long tuples in PatternLocalRewriter

Youssef1313 opened the pull request:
Fixes #47878

To explain the changes:

loweredInput.Arguments can have at **max** 8 arguments, with the 8th argument being another tuple. This is due to the available `ValueTuple`s in the runtime:

![image](https://user-images.githubusercontent.com/31348972/93796845-b2210580-fc3b-11ea-9b08-f1b103261c15.png)

I fixed the loop to have the exact number of items (which can be greater than 8), and introduced a static local function which gets the BoundExpression for the corresponding index.

1. When `index < 7`, just return `Arguments[index]`.
2. When `index >= 7`, gets the corresponding `TRest` tuple, and apply the same logic for `index - 7`.

Youssef1313 commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:
```diff
@@ -488,7 +488,7 @@ bool usesOriginalInput(BoundDecisionDagNode node)
                 out BoundExpression savedInputExpression)
             {
-                int count = loweredInput.Arguments.Length;
+                int count = loweredInput.Type.TupleElements.Length;
```
For tuple containing > 7 elements (10 elements for example), the previous value of count will be 8 (the TRest position) instead of 10, causing the loop after a few lines to run a fewer iteration than needed.

jcouv replied:
nit: consider extracting `loweredInput.Type.TupleElements` as a local (also used on line 498), then remove the `count` local.

Youssef1313 commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:
```diff
@@ -497,7 +497,7 @@ bool usesOriginalInput(BoundDecisionDagNode node)
                     var field = loweredInput.Type.TupleElements[i].CorrespondingTupleField;
                     Debug.Assert(field != null);
-                    var expr = loweredInput.Arguments[i];
+                    var expr = getExpression(i, loweredInput);
```
Here, we can't just access Arguments[i] for two reasons:

1. If i == 8, the result expression will be a tuple, this is not intended.
2. If i > 8, this will throw an exception due to the other fix I made.

Youssef1313 commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:
```diff
@@ -512,6 +512,25 @@ bool usesOriginalInput(BoundDecisionDagNode node)
+                        }
+                        else
+                        {
+                            // ValueTupleRestIndex should always be another tuple. Hence, it should always have BoundObjectCreationExpression.
+                            Debug.Assert(false, "This should be unreachable.");
```
Is my assumption here correct?

333fred replied:
I believe so. Consider `throw ExceptionUtilities.Unreachable` instead of an assert.

---
In reply to: [492204839](https://github.com/dotnet/roslyn/pull/47893#discussion_r492204839) [](ancestors = 492204839)

333fred commented:
@gafter if you are available, we'd appreciate a review :smile:

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs:
```diff
@@ -2949,6 +2949,42 @@ public static void Main()
+{   
+    public static void Main()
+    {
+        int x1 = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1) switch
```
>1, 1, 1, 1, 1, 1, 1, 1, 1, 1 [](start = 18, length = 28)

Consider making these numbers increment (ie `1, 2, 3, ...`) for better readability of the test.

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs:
```diff
@@ -2949,6 +2949,42 @@ public static void Main()
+{   
+    public static void Main()
+    {
+        int x1 = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1) switch
```
Please split out this test and verify the IL. I don't think we need to verify IL on all of these variations, but it would be good to have 1 of them verified.

Youssef1313 replied:
@333fred I feel it's safer to verify the IL for different cases because in the past, the compiler had different behavior between the 8 elements case, and the > 8 elements case.

I'd at least verify the IL for 2 cases:

- 8 elements
- Greater than 8 elements

333fred commented:
Done review pass (commit 3).

333fred commented:
Also, it looks like another test is failing with this change.

Youssef1313 commented:
> 
> 
> Also, it looks like another test is failing with this change.

I'll give it a look and try to fix that with the test changes you suggested.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:
```diff
@@ -512,6 +512,25 @@ bool usesOriginalInput(BoundDecisionDagNode node)
 
                 return rewrittenDag;
 
+                static BoundExpression getExpression(int index, BoundObjectCreationExpression expression)
```
nit: Consider folding this logic into the above loop, so that this gets repeated fewer times. This is an uncommon scenario, so readability should come first (try and see what it looks like).

Youssef1313 replied:
@jcouv Based on https://github.com/dotnet/roslyn/pull/47893#issuecomment-696386895, this method may be needed in another place. In such a case, it would make more sense to be extracted to an extension method somewhere.

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs:
```diff
@@ -2949,6 +2949,42 @@ public static void Main()
+    {
+        int x1 = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1) switch
+        {
+            (1, 1, 1, 1, 1, 1, 1, 1, 1, 1) => 1,
```
The `11` doesn't seem to line up with every 7-th position.

Youssef1313 replied:
@jcouv I didn't get this.

jcouv replied:
Seems like there was a display bug in the code review tool I use. It displayed `(1, 1, 1, 1, 1, 1, 1, 11, 1, 1, 1, 1, 1, 11, ...)`.
Sorry about that!

jcouv reviewed (commented):
Done with review pass (iteration 3). Minor comments only

Youssef1313 commented:
> 
> 
> Also, it looks like another test is failing with this change.

With the change I introduced, `GetArgumentRefKind` method is now being called with indices > 7, which sounds correct to me. It just needs a similar logic to the static local function `getExpression` that I introduced. One small problem is where to define this logic to avoid duplicating it in two places.

Another problem is that I'm unsure of "who" calls GetArgumentRefKind? is it used for things other than tuples (i.e. normal method parameters) in which case things may get a bit more complex. Also I spotted an assertion that is now passing just "by luck":

```csharp
        private void EmitArguments(ImmutableArray<BoundExpression> arguments, ImmutableArray<ParameterSymbol> parameters, ImmutableArray<RefKind> argRefKindsOpt)
        {
            // We might have an extra argument for the __arglist() of a varargs method.
            Debug.Assert(arguments.Length == parameters.Length || arguments.Length == parameters.Length + 1, "argument count must match parameter count");
            Debug.Assert(parameters.All(p => p.RefKind == RefKind.None) || !argRefKindsOpt.IsDefault, "there are nontrivial parameters, so we must have argRefKinds");
            Debug.Assert(argRefKindsOpt.IsDefault || argRefKindsOpt.Length == arguments.Length, "if we have argRefKinds, we should have one for each argument");

            for (int i = 0; i < arguments.Length; i++)
            {
                RefKind argRefKind = GetArgumentRefKind(arguments, parameters, argRefKindsOpt, i);
                EmitArgument(arguments[i], argRefKind);
            }
        }
```

This assertion:

```csharp
            // We might have an extra argument for the __arglist() of a varargs method.
            Debug.Assert(arguments.Length == parameters.Length || arguments.Length == parameters.Length + 1, "argument count must match parameter count");
```

Assumes a max difference of 1 between arguments length and parameters length.

Because the corresponding test uses a tuple of 9 elements, and arguments is 8 (where the 8th argument is a tuple itself holding the 8th and 9th values). the assertion passes. But if it's a tuple of 10 elements. arguments.Length would still be 8 and the assertion will fail.

Given that, I'm now less confident with the change I made. It breaks some existing assumptions. I could just remove the assertion in `EmitArguments`, but I'm afraid this breaks something.

333fred commented:
> One small problem is where to define this logic to avoid duplicating it in two places.

Consider doing this as a static method on `TupleTypeSymbol`. I'll take a look at the code in EmitArguments a bit more.

Youssef1313 commented:
> 
> 
> > One small problem is where to define this logic to avoid duplicating it in two places.
> 
> Consider doing this as a static method on `TupleTypeSymbol`. I'll take a look at the code in EmitArguments a bit more.

The existing code needs to get a `BoundExpression` from `BoundObjectCreationExpression` based on the index. I'm not sure if I can achieve the same if the extension method is on TupleTypeSymbol.

I found a class named `BoundExpressionExtensions`, looks like a good place for this extension method.

For the failing assertions, I think I may surround them with an if condition so they are executed for non-tuple parameters only.

333fred commented:
```cs
            // We might have an extra argument for the __arglist() of a varargs method.
            Debug.Assert(arguments.Length == parameters.Length || arguments.Length == parameters.Length + 1, "argument count must match parameter count");
```

Perhaps the right thing to do is break this assert out into a conditional method that can verify the right thing. It would verify that:
* `arguments.Length == parameters.Length` in the general case (and the last parameter is not an `__arglist`)
* `arguments.Length == parameters.Length + 1` when the last parameter is an `__arglist`
* The tuple logic as appropriate.

You'll have to surround such a method if `#if DEBUG` blocks for now, since we have a bug with `Conditional` that isn't going to be fixed in the toolset until RC2 (iirc).

Youssef1313 commented:
Actually I think I was incorrect when I said:

> With the change I introduced, GetArgumentRefKind method is now being called with indices > 7, which sounds correct to me.

This isn't correct. ValueTuple constructor have at most 8 arguments. Even if the elements are greater than that, it doesn't make sense to call ArgumentRefKind with index > 7. So `GetArgumentRefKind` should be kept as it is with the assertion in it.

The callers & caller's assertion needs some tweaks.

RikkiGibson commented:
GitHub keeps presenting my name at the top of the completion list, then the "remove all assignees" button pops in..

Youssef1313 commented:
One thing to notice, when this was working in past, the generated code was using pointers in an unsafe context.

The change in this PR tries to make the code works again, but in non-unsafe fashion.

But I can't get it to work correctly.

@RikkiGibson @jcouv Are you aware of any PRs in the past that was around making the generated code non-unsafe?

RikkiGibson commented:
@Youssef1313 I messaged you on gitter, let's figure out the answer to that and how to move forward.

RikkiGibson commented:
Summary of the situation:

Pattern matching of tuples of more than 8 elements (in this scenario at least) [has not worked](https://sharplab.io/#v2:EYLgJgpgtg9gzgWgE4QMYyWOAaALiJAVwDtswQBqAHwAEAmARgFgAoGgZgAJ7OBhTgN6tOI7nTqDho6TNEBLYrk4APTgF5OACgbZOOvbv1HDJvQEpOcAO5zcqABZTZQlrNnbTxg968MLagD4DJzcRAH11IIQdEOkAXwBuWNEaBgBOTWUzJNdRONY4oA=) for some time. At least not since the 16.5 time frame. For a while, pattern matching tuples with exactly 8 elements [has worked](https://sharplab.io/#v2:EYLgJgpgtg9gzgWgE4QMYyWOAaALiJAVwDtswQBqAHwAEAmARgFgAoGgZgAJ7OBhTgN6tOI7nTqDho6TNEBLYrk4APTgF5OACgbZOOvbv1HDhgJSc4Adzm5UACymyhLWbO0mDn41/NqAfAaOriIA+uoBCDpB0gC+ANzRojQMAJyayqYJLqIxrDFAA===), but later the semantics of the emitted code were [broken](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8ABAJgEYBYAKGIGYACMhgYQYG8aHunTSOuPIcJ4BLTDAYAPBgF4GACnJIGy1SrWaNGgJQMoAd1EwAxgAtBIztREil29Y61O9sgHzrLt7gH05HgFplLyEAXwBuEJ5icgBOBSkdSJseUJpQoA=).

Perhaps we should just look at the up-to-date implementation and try to fix it to work with indefinitely many elements, perhaps using a "too many locals" or similar diagnostic if we hit a meaningful limit.

Youssef1313 commented on src/Compilers/CSharp/Test/Semantic/Semantics/SwitchTests.cs:
```diff
@@ -2949,6 +2949,234 @@ public static void Main()
+        }
+
+        [Fact, WorkItem(47878, "https://github.com/dotnet/roslyn/issues/47878")]
+        public void VerifyIL_8ElementsTuple()
```
Is the IL for this test and the next test correct?

Youssef1313 commented:
@RikkiGibson Looks like it should be working now. But my changes removed a whole array that used to exist, which makes me a little bit less confident. Can this change break some scenario that is currently not tested?

RikkiGibson commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:
```diff
@@ -488,30 +488,47 @@ bool usesOriginalInput(BoundDecisionDagNode node)
                 var rewrittenDag = decisionDag.Rewrite(makeReplacement);
                 savedInputExpression = loweredInput.Update(
-                    loweredInput.Constructor, arguments: newArguments.ToImmutableAndFree(), loweredInput.ArgumentNamesOpt, loweredInput.ArgumentRefKindsOpt,
+                    loweredInput.Constructor, arguments: loweredInput.Arguments, loweredInput.ArgumentNamesOpt, loweredInput.ArgumentRefKindsOpt,
```
I'm confused by the change to get rid of the `newArguments` array. I thought that would have undesired effects, particularly related to reuse of user temps or side effects.

Youssef1313 replied:
@RikkiGibson I was expecting to see some test failures related to this too, but I don't know what scenarios exactly that should fail. Do you know any scenarios that should fail to add tests for them?

RikkiGibson replied:
I tried the following scenario to induce a failure but it continued to have the expected side-effects in your branch:

```cs
        [Fact, WorkItem(47878, "https://github.com/dotnet/roslyn/issues/47878")]
        public void VerifyIL_8ElementsTuple_SideEffects()
        {
            var text = @"
using System;
public class C
{
    public static void Main()
    {
        int x = (M(1), M(2), M(3), M(4), M(5), M(6), M(7), M(8)) switch
        {
            (1, 2, 3, 4, 5, 6, 7, 7) => 1,
            (1, 2, 3, 4, 5, 6, 7, 8) => 1,
            _ => -1,
        };

        Console.WriteLine(x);
    }

    static int M(int x)
    {
        Console.Write(x);
        return x;
    }
}
";
            CompileAndVerify(text, expectedOutput: "123456781");
        }
```

I think the next thing I would try would be to make many of the tuple components local variables and see if the lowering code reuses the local variables properly instead of introducing new ones.

I'll try to debug in more detail soon.

RikkiGibson replied:
Here's the scenario where we have unexpected side effects:

```cs
        [Fact, WorkItem(47878, "https://github.com/dotnet/roslyn/issues/47878")]
        public void VerifyIL_8ElementsTuple_SideEffects_02()
        {
            var text = @"
using System;
using System.Runtime.CompilerServices;
public class C
{
    public static void Main()
    {
        try
        {
            int x = (M(1), M(2), M(3), M(4), M(5), M(6), M(7), M(8)) switch
            {
                (1, 2, 3, 4, 5, 6, 7, 7) => 1,
            };

            Console.WriteLine(x);
        }
        catch (SwitchExpressionException ex)
        {
            Console.Write(""💥"");
        }
    }

    static int M(int x)
    {
        Console.Write(x);
        return x;
    }
}
namespace System.Runtime.CompilerServices
{
    public class SwitchExpressionException : InvalidOperationException
    {
        public SwitchExpressionException() {}
        public SwitchExpressionException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
        public object UnmatchedValue { get; }
    }
}
";
            CompileAndVerify(text, expectedOutput: "12345678💥");
        }
```

With your change, the actual output is `1234567812345678💥`. Let's revert that part of the change.

Youssef1313 replied:
@RikkiGibson That will get us to the assertion failure too. I'm still not sure what is the appropriate fix. I'll add this test and give it a try again, but not sure if I would be able to get the correct fix.

RikkiGibson replied:
I think the problem is that when we "unwrap" the tuple elements as is done in this PR, the length of the arguments no longer matches the length of the arguments ot the top-level object creation node. i.e. `loweredInput.Arguments.Length` should never be more than 8, and tuples of size 8 or larger must nest their arguments in additional object creation nodes.

I can try and look a bit to see how to do this--the fix isn't coming to mind instantly. But to get the behavior of the scenario fixed in 16.8 we may want to take #48028 and then optimize it later.

RikkiGibson commented:
I will move this back to draft status. Feel free to mark it ready for review again if you want us to take another look later.

Youssef1313 commented:
Closing this stale PR.
